pkg/chunk/purger/purger.go

```diff
@@ -296,9 +323,15 @@ func (p *Purger) worker() {
 	}
 }
 
-func (p *Purger) executePlan(userID, requestID string, planNo int, logger log.Logger) error {
+func (p *Purger) executePlan(userID, requestID string, planNo int, logger log.Logger) (err error) {
 	logger = log.With(logger, "plan_no", planNo)
 
+	defer func() {
+		if err != nil {
+			p.inProcessRequests.setFailedRequestForUser(userID)
```

Comment:
Instead of using `defer`, it would be clearer to explicitly call `setFailedRequestForUser` in branches where it is needed.

I don't think all errors need to be retried (eg. when parsing of chunkID fails).

Same comment applies to `buildDeletePlan`.

Reply:
I could not point out any acceptable failures which is why I considered all failures must be retried.
Having a non-parseable chunk id is something which needs someone's intervention because even queries would start failing.

We can always change this if we see cases where we are unnecessarily retrying a non-recoverable and ingorable case. Missing or ignoring a case like non-parseable chunkID could be something which is not expected to happen.

What do you think?

Reply:
My thinking was more that I don't see a point in repeating failed request, if we expect the same failure (eg. that chunk parsing -- I would expect that to just fail over and over, or not?).

Reply:
Yes, it would fail over and over until someone manually intervenes and fixes the problem. The fix would be either fixing the chunk or removing it. Same would happen when someone queries data touching that chunk.
It could a problem which needs to be noticed by someone and not everyone would go and check the logs for errors like this.

Reply:
I don't think wasting resources on jobs we know will fail again and again is a good approach. It should raise an alert instead.

Reply:
It would alert only if it fails. We have alerts setup for oldest pending request.
We also have a dashboard for showing failures in processing of request. Constant failures would be noticed but a single failure once in a while is not noticeable.